Report readers documentation (#78)
BlueBrain/libsonata: README.rst

   >>> selection = population.connecting_edges([1, 2, 3], [4, 5, 6])

Reports
~~~~~~~

SpikeReader
+++++++++++

.. code-block:: pycon

   >>> import libsonata

   >>> spikes = libsonata.SpikeReader('path/to/H5/file')

   # list populations
   >>> spikes.get_populations_names()

   # open population
   >>> population = spikes['<name>']

SpikePopulation
+++++++++++++++

.. code-block:: pycon

   # get all spikes [(node_id, timestep)]
   >>> population.get()
   [(5, 0.1), (2, 0.2), (3, 0.3), (2, 0.7), (3, 1.3)]

   # get all spikes betwen tstart and tstop

>    >>> population.get(0.2, 1.0)

Reviewer: Maybe better to advertise the named argument instead (less error prone) ?

```
population.get(tstart=0.2, tstop=1.0)
```
Also as I said in : 
https://github.com/BlueBrain/libsonata/pull/96

There is a bunch of implicit behavior we need to specify here: 

```
population.get(tstart=-1, tstop=0.5)
```
Negative values for times stand for minimum report time for tstart and maximum report time for tstop.
Also `node_ids=[]` returns all node_ids.
So : 
```
population.get() == population.get(tstart=-1, tstop=-1, node_ids=[])
```

Author: > Maybe better to advertise the named argument instead (less error prone) ?
> 
> ```
> population.get(tstart=0.2, tstop=1.0)
> ```
I totally agree, this way is much clearer 

> Also as I said in :
> #96
> 
> There is a bunch of implicit behavior we need to specify here:
> 
> ```
> population.get(tstart=-1, tstop=0.5)
> ```
> 
> Negative values for times stand for minimum report time for tstart and maximum report time for tstop.
> Also `node_ids=[]` returns all node_ids.
> So :
> 
> ```
> population.get() == population.get(tstart=-1, tstop=-1, node_ids=[])
> ```

I am not sure if explaining this in the readme is the best place, I think we should keep this simple here and maybe extend the documentation for the more complicated cases somewhere else?

Reviewer: For snap we have something like this : 
https://bluebrainsnap.readthedocs.io/en/latest/submodules/bluepysnap.nodes.html#bluepysnap.nodes.NodePopulation.ids

So all the possible inputs are listed here and this draw directly from the docstrings. 

What is cool about that is we also have this doc when you do in your python console: 
```
help(NodePopulation.ids)
```
So I am not sure what s the best place for this can you check with @haleepfl  ?

Reviewer 2: This type of documentation should definitely be in the docstrings. It is best to have the documentation close to the source, and gives it the best chance of staying up-to-date if the behavior changes in the future. This is also applicable for the C++ doxygen strings.

   [(2, 0.2), (3, 0.3), (2, 0.7)]

   # get spikes attribute sorting (by_time, by_id, none)
   >>> population.sorting
   'by_time'

> 

Reviewer: If you want to provide a dataframe example : 

```
data = population.get()
dataframe = pandas.DataFrame(data=data, columns=["ids", "times"]).set_index("times")
```

   >>> data_frame = population_somas.get(node_ids=[13, 14], tstart=0.8, tstop=1.0)

   # get the data values
   >>> data_frame.data
   [[13.8, 14.8], [13.9, 14.9]]

   # get the list of timesteps
   >>> data_frame.times
   [0.8, 0.9]

   # get the list of node ids
   >>> data_frame.ids
   [13, 14]

Pandas can be used to create a dataframe using the data, ids and times lists

.. code-block:: pycon

   >>> import pandas

   df = pandas.DataFrame(data_frame.data, columns=data_frame.ids, index=data_frame.times)
   print(df)
          13    14
   0.8  13.8  14.8
   0.9  13.9  14.9

ElementReportReader
+++++++++++++++++++

.. code-block:: pycon

   >>> elements = libsonata.ElementReportReader('path/to/H5/file')

   # list populations
   >>> elements.get_populations_names()

   # open population
   >>> population_elements = elements['<name>']

ElementReportPopulation
+++++++++++++++++++++++

.. code-block:: pycon

   # get times (tstart, tstop, dt)
   >>> population_elements.times
   (0.0, 4.0, 0.2)

   # get the DataFrame of the node_id values for the timesteps between tstart and tstop
   >>> data_frame = population_elements.get(node_ids=[13, 14], tstart=0.8, tstop=1.0)

>    # get the data values (map of [node_id, element_id] -> values[timesteps]

Reviewer: not anymore :
```suggestion
   # get the data values (list of list of floats with data[time_index][element_index])
```

   >>> data_frame.data
   [[46.0, 46.1, 46.2, 46.3, 46.4, 46.5, 46.6, 46.7, 46.8, 46.9], [56.0, 56.1, 56.2, 56.3, 56.4, 56.5, 56.6, 56.7, 56.8, 56.9]]

   # get the list of timesteps
   >>> data_frame.times
   [0.8, 1.0]

   # get the list of (node id, element_id)
   >>> data_frame.ids
   [(13, 30), (13, 30), (13, 31), (13, 31), (13, 32), (14, 32), (14, 33), (14, 33), (14, 34), (14, 34)]

The same way than with soma reports, pandas can be used for better usability

.. code-block:: pycon

   >>> import pandas

>    df = pandas.DataFrame(data_frame.data, columns=data_frame.ids, index=data_frame.times)

Reviewer: Even if what you proposed here is perfectly valid, I think it is better to use : 

```suggestion
   df = pandas.DataFrame(data_frame.data, columns=pandas.MultiIndex.from_tuple(data_frame.ids), index=data_frame.times)
```
Doing `columns=data_frame.ids` would give you a standard index instead of a hierarchical one. 
https://pandas.pydata.org/pandas-docs/stable/user_guide/advanced.html

With multiindex, you can retrieve more easily all the compartments of  node ids.

Reviewer: Sorry this is `from_tuples` with a `s`

>    df = pandas.DataFrame(data_frame.data, columns=pandas.MultiIndex.from_tuple(data_frame.ids), index=data_frame.times)

Reviewer: sorry I did a typo :
```suggestion
   df = pandas.DataFrame(data_frame.data, columns=pandas.MultiIndex.from_tuples(data_frame.ids), index=data_frame.times)
```

   print(df)

>         (13, 30)   (13, 30)   (13, 31)   (13, 31)   (13, 32)  (14, 32)   (14, 33)   (14, 33)   (14, 34)   (14, 34)

Reviewer: this will change to : 
```
          13                                                                                 14                                            
          30               30               31               31                32    32               33               33               34               34
0.8  46.0  46.099998  46.200001  46.299999  46.400002  46.5  46.599998  46.700001  46.799999  46.900002
1.0  56.0  56.099998  56.200001  56.299999  56.400002  56.5  56.599998  56.700001  56.799999  56.900002

```

Reviewer: That s with the MultiIndex

Reviewer: rahhh git hub changed the formatting ... 

```
       14                                                13                                            
       32         33         33         34         34    30         30         31         31         32
0.8  46.5  46.599998  46.700001  46.799999  46.900002  46.0  46.099998  46.200001  46.299999  46.400002
1.0  56.5  56.599998  56.700001  56.799999  56.900002  56.0  56.099998  56.200001  56.299999  56.400002

```

Author: haha dont worry, i will run the test and copy the output :D
